Refactor and remove legacy APIs/testing internals.

stuartarchibald opened the pull request:
This PR removes the following (and does some minimal refactoring in the process):
* `numba.core.compiler.compile_isolated` this function was used a lot in Numba's test suite and exists from a time where the compiler was less isolated than it is now. The function itself also relies on "nested contexts" within the same target which has been discovered to cause numerous issues in practice.
* `numba.tests.support.TestCase::run_nullary_func` this method isn't really used very much in the test suite or needed any more, refactoring to remove it was trivial.
* `numba.tests.support.CompilationCache`, this was used in the test suite from a time when the Numba dispatchers were less well isolated and `compile_isolated` needed somewhere to cache results, this isn't needed any more since `compile_isolated` is removed.
* "nested context" as a concept from the `numba.core.registry.CPUTarget`, this was used a lot in conjunction with the now removed `compile_isolated`, it has a history of causing hard to debug problems.

Notes for the reviewer:
1. These patches are reasonably isolated to one file being changed per commit.
2. Patch b91050302 fixes a bug in `_typeof.cpp` that was hidden by `compile_isolated`, it is therefore less "easy" to review.
3. Patch 834ec6b53 also includes a minor refactor to `test_svml.py` to better use existing testing infrastructure to achieve isolation.

stuartarchibald commented on numba/tests/test_dataflow.py:
```diff
@@ -27,21 +20,25 @@ def assignments2(a):
 
 # Use cases for issue #503
 
+@njit((types.intp, types.intp))
```
NOTE: this is going to violate the assertions made in #9309.

guilhermeleobas commented:
I'll review from [1305229](https://github.com/numba/numba/pull/9330/commits/13052298a121f5278801bd5b64f65be7014c3ba4) to [7ea64ab](https://github.com/numba/numba/pull/9330/commits/7ea64aba0dcb6164cb3e6ee39797f2095a063d63)

sklam commented on numba/tests/test_record_dtype.py:
```diff
@@ -1047,12 +1044,11 @@ def test_py_argument_charseq(self):
-                                         rectype.typeof('n')))
-        cfunc = cres.entry_point
+        sig = (rectype[::1], types.intp, types.Bytes(types.uint8, 1, 'C'))
+        cfunc = njit(sig)(pyfunc)
 
         for i in range(self.refsample1d.size):
-            chars = "{0}".format(hex(i + 10))
+            chars = bytes("{0}".format(hex(i + 10)), 'UTF-8')
```
This change defeats the intention of the test, which is to test `CharSeq` as argument. Suggest revert then change 
 
```suggestion
        sig = (rectype[::1], types.intp, rectype.typeof('n'))
        cfunc = njit(sig)(pyfunc).overloads[sig].entry_point

        for i in range(self.refsample1d.size):
            chars = "{0}".format(hex(i + 10))
```

stuartarchibald replied:
Thanks, fixed in ce9f24e.

sklam commented on numba/tests/test_record_dtype.py:
```diff
@@ -1063,16 +1059,15 @@ def test_py_argument_char_seq_near_overflow(self):
-        cres = compile_isolated(pyfunc, (rectype[:], types.intp,
-                                         rectype.typeof('n')))
-        cfunc = cres.entry_point
+        sig = (rectype[::1], types.intp, types.Bytes(types.uint8, 1, 'C'))
+        cfunc = njit(sig)(pyfunc)
```
same here
```suggestion
        sig = (rectype[::1], types.intp, rectype.typeof('n'))
        cfunc = njit(sig)(pyfunc).overloads[sig].entry_point
```

stuartarchibald replied:
Thanks, fixed in ce9f24e.

sklam commented on numba/tests/test_record_dtype.py:
```diff
@@ -1063,16 +1059,15 @@ def test_py_argument_char_seq_near_overflow(self):
                          recordwithcharseq['n'].itemsize)
 
-        cfunc(self.nbsample1d, 0, cs_near_overflow)
+        cfunc(self.nbsample1d, 0, bytes(cs_near_overflow, 'UTF-8'))
```
```suggestion
        cfunc(self.nbsample1d, 0, cs_near_overflow)
```

stuartarchibald replied:
Thanks, fixed in ce9f24e.

sklam commented on numba/tests/test_record_dtype.py:
```diff
@@ -1084,14 +1079,13 @@ def test_py_argument_char_seq_truncate(self):
-        cres = compile_isolated(pyfunc, (rectype[:], types.intp,
-                                         rectype.typeof('n')))
-        cfunc = cres.entry_point
+        sig = (rectype[::1], types.intp, types.Bytes(types.uint8, 1, 'C'))
+        cfunc = njit(sig)(pyfunc)
```
same here:

```suggestion
        sig = (rectype[::1], types.intp, rectype.typeof('n'))
        cfunc = njit(sig)(pyfunc).overloads[sig].entry_point
```

stuartarchibald replied:
Thanks, fixed in ce9f24e.

sklam commented on numba/tests/test_record_dtype.py:
```diff
@@ -1084,14 +1079,13 @@ def test_py_argument_char_seq_truncate(self):
 
         pyfunc(self.refsample1d, 1, cs_overflowed)
-        cfunc(self.nbsample1d, 1, cs_overflowed)
+        cfunc(self.nbsample1d, 1, bytes(cs_overflowed, 'UTF-8'))
```
```suggestion
        cfunc(self.nbsample1d, 1, cs_overflowed)
```

stuartarchibald replied:
Thanks, fixed in ce9f24e.

sklam commented on numba/tests/test_svml.py:
```diff
@@ -345,96 +328,81 @@ def check(self, pyfunc, *args, **kwargs):
-    def test_svml(self):
+    def test_scalar_context_result(self):
+        # checks result for test_scalar_context_asm
+        self.check(math_sin_scalar, 7., what="result")
+        self.check(math_sin_scalar, 7., what="result")
```
why run twice?

stuartarchibald replied:
Copy paste error I guess, fixed in 5bd7254.

sklam reviewed (commented):
My review from 7ea64ab to 67c6762d2 (inclusive)

sklam reviewed (commented):
Here's my review from 67c6762d22169b90dc6d359325e2f666d709a023 to e949711d674664ef09eec48d0ca32e03089a5372.

The changes in this range is almost mechanical removal and is easy to review.

sklam commented on numba/tests/test_return_values.py:
```diff
@@ -31,13 +30,12 @@ class TestReturnValues(unittest.TestCase):
             self.assertEqual(result, abs)
         else:
-            result = cfunc()
+            cfunc()
```
Dead code because the compiler would have failed.

sklam replied:
```suggestion
            self.fail("unexpected to compile successfuly")
```

stuartarchibald replied:
Thanks, fixed in 90ffc0544d1.

sklam commented on numba/tests/test_return_values.py:
```diff
@@ -61,13 +58,12 @@ def test_pyobj_func_npm(self):
             self.assertEqual(result, math.floor)
         else:
-            result = cfunc()
+            cfunc()
```
Dead code because the compiler would have failed.

stuartarchibald replied:
Thanks, fixed in 90ffc0544d1.

sklam commented on numba/tests/test_return_values.py:
```diff
@@ -46,13 +44,12 @@ def test_nopython_func_npm(self):
             self.assertEqual(result, open)
         else:
-            result = cfunc()
+            cfunc()
```
```suggestion
            self.fail("unexpected to compile successfuly")
```

stuartarchibald replied:
Thanks, fixed in 90ffc0544d1.

sklam reviewed (commented):
I reviewed the rest of the files:

stuartarchibald commented on numba/tests/test_auto_constants.py:
```diff
@@ -4,30 +4,25 @@
 
     def test_sys_constant(self):
-        def pyfunc():
+        def f():
```
Thanks, fixed in 4ac0cec.

stuartarchibald commented on numba/tests/test_operators.py:
```diff
@@ -770,11 +761,17 @@ def test_truediv_complex_npm(self):
+            cres(4j, 2j)
+
+        msg = "unsupported operand type(s) for %"
+        self.assertIn(msg, str(raises.exception))
```
Thanks, fixed in 62576f8.

stuartarchibald commented on docs/upcoming_changes/9330.change.rst:
```diff
@@ -0,0 +1,16 @@
+* ``numba.core.compiler.compile_isolated`` is removed.
+* ``numba.tests.support.TestCase::run_nullary_func`` is removed.
+* ``numba.tests.support.CompilationCache`` is removed.
+* "nested context" as a concept is removed from
```
Good idea, done in cd3d356, see what you think?

stuartarchibald commented on numba/tests/test_array_iterators.py:
```diff
@@ -281,29 +274,40 @@ def test_array_flat_3d(self):
+            cres = cfunc.overloads[(arrty,)]
+            got = cres.entry_point(arr)
+            expected = cfunc.py_func(arr)
+            np.testing.assert_allclose(expected, got)
```
Thanks, fixed in 47fc3902.

stuartarchibald commented on numba/tests/test_array_iterators.py:
```diff
@@ -154,17 +149,15 @@ def check_array_flat(self, arr, arrty=None):
+        cfunc = njit((arrty,))(func)
         self.assertPreciseEqual(cfunc(arr), func(arr))
 
     def check_array_flat_sum(self, arr, arrty):
```
Thanks, fixed in 3cd54ee.

stuartarchibald commented on numba/tests/test_tuples.py:
```diff
@@ -176,47 +175,44 @@ def test_hetero_tuple(self):
+        cfunc = njit((tuple_type,))(tuple_first)
+        with self.assertRaises(TypeError) as raises:
+            cfunc((4, 5, 6))
+        self.assertIn("No matching definition", str(raises.exception),)
```
Thanks, fixed in 271e2e1b.

stuartarchibald commented on numba/tests/test_array_manipulation.py:
```diff
@@ -312,13 +308,12 @@ def test_array_transpose_axes(self):
         def run(pyfunc, arr, axes):
-            cres = self.ccache.compile(pyfunc, (typeof(arr), typeof(axes)))
-            return cres.entry_point(arr, axes)
+            return pyfunc(arr, axes)
```
Thanks, fixed in e0c849d.

stuartarchibald commented:
Many thanks for the review @sklam. I think I've addressed all the comments and have added a SHA to each one to hopefully make it easier to see the change that implements the fix.

I think there's two changes made that do not correspond to a requested change:
*  b22f8dd is fixing up parfors testing to look for the gufunc assembly output that matches the compilation of interest opposed to checking all outputs (there can now be a few as the compilation context is shared). This change is require to make the parfors tests pass.
* 7ef939a is fixing up against #9309.

There's very little chance of public CI being able to successfully run with this diff as too many tests have changed and the time limits will be exceeded as a result of running them all, we'll have to rely on the build farm.

stuartarchibald commented:
Speculatively testing 7ef939a on the buildfarm with ID: `numba_smoketest_cpu_yaml_186`.

sklam commented on docs/upcoming_changes/9330.change.rst:
```diff
@@ -0,0 +1,23 @@
+API in ``numba.core.target_extension`` to extend Numba support to
+custom/synthetic hardware) should note that the same can be deleted from
+target extension implementations of ``numba.core.descriptor.TargetDescriptor``
+if it present. i.e. the ``nested_context`` method and associated implementation
```
actually, do you even need "if it present" here. the sentence is already long

stuartarchibald replied:
I wanted the `if` in there to make it so that users are unconcerned if their implementation doesn't have one as it's not a requirement. For example the CUDA target does not have a `nested_context`:
https://github.com/numba/numba/blob/688c5141e96eb81dc365a864419c1141b2949333/numba/cuda/descriptor.py#L10-L30

stuartarchibald commented:
> I have reviewed everything. It's just pending the one spelling fix and buildfarm to complete

Many thanks @sklam, spelling/typo fix is in 99a7655. Conflicts against `main` (688c5141) resolved in ad5f660.

stuartarchibald commented:
> Speculatively testing [7ef939a](https://github.com/numba/numba/commit/7ef939a17809cd1ca0c10dab8f458a465d8166d1) on the buildfarm with ID: `numba_smoketest_cpu_yaml_186`.

This passed.

stuartarchibald commented:
@sklam 086b212 should resolve the merge conflict with `main`, this branch should now be the same as #9349.

sklam reviewed (approved):
LGTM

RE: https://github.com/numba/numba/pull/9330#issuecomment-1850319931

indeed it matches

sklam commented:
Not going to wait for CI
